### ISSUE TITLE
[OS X] Improve client build and packaging instructions

### ISSUE THREAD

**Issue by Delta-38** [owncloud/client#5594]
Hello, we are trying to build the Mac client for owncloud (after successfully building the windows and Linux clients) however, even following the instructions to the letter, and trying on several different development Macs at our office, we cannot get it to complete. 

We are trying to build on a Mac with OSX 10.12 and have tried both building with qt4 (with the corresponding option on ) and with qt5 (trie 5.6, 5.7 and about to try with 5.8...)

We have faced the following issues: 

1)The instructions for MacOSX aren't very clear as to the actual folder structure that has to be attained relative to the client and qtkeychain projects, there is frequent reference to install-prefix dirs, but it is unclear as to wheter the qtkeychain cmake and the client cmake should aim at the same folder or not. Other references to build-dir, install-dir are often unclear as to what they refer to. 
2) Is there a way to use the installed version of Qt (The ones coming with the Qt installer rather than the brew version? ) This is a recent iMac and after 140 minutes it is still stuck in the brew install qt5 ... step... (The comments related to other sections of the install process have already been run on other machines here)

3) When we run cmake on the client it comments not finding iconv.h we tried solving this pointing CMAKE_PREFIX_PATH to the Qt lib folder, but the problem reemerged in the create_mac.sh phase.
4)We also get an error in the creation of the package where it says macosx.pkgproj is not a packages project file. (On the machine that got furthest with the build attempts).
4.1) On this machine the owncloud.app runs, but the created package won't run on other Macs, we tried with running macdeployqt over the .app but it didn't help. 
5) the documented "use QT4" cmake parameter is ignored. When used we get the warning qt5 was not found.

I understand it is a lot to ask and probably several separate comple issues, but we are stumped and we would love to get some help with the build.
We feel that the Mac build process documentation could use some love, a probably a bit less optimism. :)

Paolo Manili
 
### Client configuration
Client version:2.3.0
Operating system: OSX 10.12
OS language: Italian

**Comment by Delta-38** [comment]
Hi, we finally got it to build successfully, I am posting the steps we followed and some friendly suggestions for the documentation :) :

We eventually managed to build first the standard owncloud and then our customized version. 

We identified two separate problems that were stopping us: 
We had first to proceed to the older "brew install qtkeychain"  approach to get a usable keychain, as per this old blogpost: 
https://jose-manuel.me/2014/02/compile-latest-owncloud-client-mac-osx-mavericks/
The second thing that was unclear is that the installer cannot be created correctly if you don't run make install on the mac_build dir. Given this part was marked optional and it seemed counterintuitive to install on the development machine (As we felt it would potentially alter the environment). Once that was done, the installer produced by mac_create_build.sh worked as expected and a usable installer was produced.

We also had to call packages manually as we coulnd't figure out the right parameters from the error using the .sh version: 
This worked for us:  run from /pathtoowncloud/mac_build/
  581  /usr/local/bin/packagesbuild -F ../install/ admin/osx/macosx.pkgproj 


So I guess we have to apologize for misunderstanding part of the documentation.
However, we would suggest the following changes to the documentation: 

Specify absolute path (With some example path given) for all operations pertaining the mix and match of owncloud and qtkeychain otherwise it is terribly easy to mixup what goes where and the number of libs not found by simple path mistake gives few clews as to where we screwed up. We still don't know, as we eventually used the brew supplied keychain. 

The documentation makes abundant use of "install" as a word and target dir, but it is not always clear if what is meant is a system install, a build dir, an include dir or what. This can probably be fixed by slightly more abundant comments. 

The make install part is "NOT" optional for a functional installer being built, and should either be mentioned in the "#installer" documentation or the build phase. 

I also feel it would probably also help if the different build processes were in consecutive sections throughout the process so Mac from scratch to installer, windows same, unix same, jumping back and forth is confusing when unfamiliar with the project and tech.

It would also help immensely if a default "will work for most of you guys" cmake command was provided and leaving advanced options for the more savvy users, whereas it is currenty the opposite as a complicated unexplained command is given:
This is provided: cmake -DCMAKE_PREFIX_PATH=/usr/local/opt/qt5 -DCMAKE_INSTALL_PREFIX=/Users/path/to/client/install/ -D_OPENSSL_LIBDIR=/usr/local/opt/openssl/lib/ -D_OPENSSL_INCLUDEDIR=/usr/local/opt/openssl/include/ -D_OPENSSL_VERSION=1.0.2a -DOPENSSL_INCLUDE_DIR=/usr/local/opt/openssl/include/ -DNO_SHIBBOLETH=1
For example I still don't know what shibboleth is, and I googled for it... :)

this is not: 
cmake -DCMAKE_BUILD_TYPE="Debug" -DCMAKE_INSTALL_PREFIX=../install .. 

TL:DR: 
Not your fault but the documentation is a bit vague :)


Thanks in advance,
Paolo Manili

**Comment by guruz** [comment]
Could you please send a pull request with documentation fixes?

:)

**Comment by settermjd** [comment]
@Delta-38 thanks for the details. I'm going to step through the process this afternoon and see what I find.

**Comment by settermjd** [comment]
It's been a while since I looked at this. But I'm looking at it this week. I'm thinking that the simplest way to resolve it going-forward is to create a shell script that can be run on macOS.

**Comment by Delta-38** [comment]
Yes I think that would be a perfectly fine solution.

2017-05-02 14:38 GMT+02:00 Matthew Setter <notifications@github.com>:

**Comment by settermjd** [comment]
I've gone through the steps and have been able to build it. But there are steps that I don't understand, which I need more information on before I can document it. And I agree that the process isn't documented correctly (and is overly complicated).

**Comment by Delta-38** [comment]
Thanks for the effort, we appreciate what you are doing and look forward to
the revised documentation. We know one does not simply walk into mordor :)

2017-05-24 15:25 GMT+02:00 Matthew Setter <notifications@github.com>:

**Comment by settermjd** [comment]
Sorry, I see that I dropped the proverbial ball on this one. Picking it up again to conclusion now.

**Comment by ArthurChiao** [comment]
hi    @all,

I've also built a client installer for MAC OSX after a trial and error process. I'd like to share some info with you.

Main references:

1. [owncloud 2.3 build guide](https://doc.owncloud.org/desktop/2.3/building.html)
1. [owncloud 2.1 build guide](https://doc.owncloud.org/desktop/2.1/building.html)
1. [Install QT with WebKit (OSX 10.12)](https://github.com/thoughtbot/capybara-webkit/wiki/Installing-Qt-and-compiling-capybara-webkit#macos-sierra-1012-el-capitan-1011-and-yosemite-1010)

----

os and software versions:

* OS: MAC OSX `10.12.6`
* xcode: `9.1` (full version, download from App Store)
* brew: `1.3.6`
* MacPorts: `2.4.0`
* client: 2.2 (customized, but i think official client code should also be ok)
* qt: `5.5.1` (get through `qmake --version`)
* qt-webkit: `?` (installed through `macport`)

-----

Problems & Solutions:

1. `qtkeychain`

    Compile and install this from source. If you encountered version mismatch problem, check out a previous tag (e.g. I fall back to v0.5.0), compile and install.

    ```
    $ git clone https://github.com/frankosterfeld/qtkeychain.git
    $ git checkout -b <br> <tag>
    $ cmake -D DCMAKE_INSTALL_PREFIX=<path to cdrive make-install dir>
    $ make && make install
    ```

1. Install QT with WebKit

    I cannot install QT with webkit by following the official docs. Later QT versions (e.g. 5.9) has dropped the WebKit support, I tried many methods, and finally get it installed by following Reference 3 in the above, using `macports` tool.

1. Mistakes in official doc

    Right: `./admin/osx/create_mac.sh <your-make-install-dir> <your build dir>`, the doc got the two paths opposite.

    Example usage:
    ```
    $ cd client-build
    $ ./admin/osx/create_mac.sh ../client-install ./
    ```

1. Dynamic library dependencies

    When packing the compiled client into an installer, the packing scripts ignores the dylib version, which may result to the installed client crashes on startup, because it references lower version libs (e.g. `libxml2.2.dylib`) in system path.

    This problem need more space to explain, a quick remedy is to modify `FindSystemLibrary()`, let it always return `None`, then re-execute the packing command.

**Comment by guruz** [comment]
https://github.com/owncloud/client/pull/6423
